File: packages/agent/agent/tools/mark_reviewed.ts
import { getDb } from "@clara-financas/db";
import { transactions } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, eq, inArray } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { requireTenantCaller, tenantIdOf } from "../lib/tenant";

/**
 * O atestado: "uma pessoa olhou isto".
 *
 * A fila de revisão junta o que a extração não fechou — sem categoria, leitura
 * de baixa confiança, sem comerciante. A conclusão mais comum de uma revisão é
 * "a leitura já estava certa", e sem registrar que alguém olhou, a fila
 * devolveria esses mesmos itens para sempre. As colunas existem desde a
 * migração 0015 e só a tela sabia escrevê-las; pela conversa — que é onde a
 * pessoa de fato confere — não havia caminho.
 *
 * Poucas linhas passam sem gate: não muda nenhum dado financeiro e é
 * reversível na frase seguinte com `reopen`. Mas o atestado em MASSA é outra
 * coisa — "marquei as 500 como revisadas" esvazia a fila inteira com uma
 * frase, e ninguém olhou 500 linhas numa frase. Acima do limiar, o cartão
 * aparece: é o mesmo argumento de escala que separa
 * `set_transaction_category` de `recategorize_transactions`. Reabrir nunca
 * exige cartão — devolver itens à fila é a direção segura.
 */
const NO_CARD_LIMIT = 20;

export default defineTool({
  description:
    "Marks entries as reviewed by the person, or reopens them with reopen=true. Use when the conclusion is 'the reading was already right' — without it the review queue keeps handing back the same entries forever. Changes no financial data. Up to 20 entries apply directly; more than that opens the approval card, because attesting in bulk empties the queue.",
  inputSchema: z.object({
    transactionIds: z.array(z.string().min(1)).min(1).max(500),
    reopen: z
      .boolean()
      .optional()
      .describe("true clears the review mark, putting the entries back in the queue."),
  }),

  approval: (ctx) => {
    const current = tenantIdOf(ctx.session.auth.current);
    if (current === undefined || current !== tenantIdOf(ctx.session.auth.initiator)) {
      return { type: "denied", reason: "A sessão não está fixada a um único usuário." };
    }
    const input = ctx.toolInput as
      | { transactionIds?: string[]; reopen?: boolean }
      | undefined;
    // Reabrir devolve à fila — direção segura, nunca pede cartão. Atestar em
    // massa esvazia a fila, e isso a pessoa decide.
    const count = input?.transactionIds?.length ?? 0;
    return input?.reopen !== true && count > NO_CARD_LIMIT ? "user-approval" : "not-applicable";
  },

  async execute(input, ctx) {
    const { tenantId, userId } = requireTenantCaller(ctx);
    const reopen = input.reopen === true;

    return forTenant(
      tenantId,
      async (tx) => {
        const touched = await tx
          .update(transactions)
          .set(
            reopen
              ? { reviewedAt: null, reviewedBy: null }
              : { reviewedAt: new Date(), reviewedBy: `human:${userId}` },
          )
          .where(
            and(
              eq(transactions.tenantId, tenantId),
              inArray(transactions.id, input.transactionIds),
              // Rascunho não se atesta: ele ainda vai passar pela decisão do
              // lote inteiro, que é uma conferência mais forte que esta.
              inArray(transactions.status, ["confirmed", "adjustment"]),
            ),
          )
          .returning({ id: transactions.id });

        const done = new Set(touched.map((row) => row.id));
        const missed = input.transactionIds.filter((id) => !done.has(id));

        return {
          [reopen ? "reopened" : "reviewed"]: touched.length,
          notFound: missed,
          ...(missed.length > 0
            ? {
                note: "Ids em notFound não existem OU ainda estão em lote proposto, onde a revisão é a decisão do lote.",
              }
            : {}),
          auditedBy: `human:${userId}`,
        };
      },
      getDb(),
    );
  },
});

File: services/agent/clara/tools/mark_reviewed.py
"""O atestado: "uma pessoa olhou isto" — porta de `agent/tools/mark_reviewed.ts`.

Não muda nenhum dado financeiro, e é reversível na frase seguinte com
`reopen=True`. A conclusão mais comum de uma revisão é "a leitura já estava
certa", e sem gravar que alguém olhou, a fila devolveria os mesmos itens para
sempre.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime

from sqlalchemy import update
from sqlalchemy.orm import Session

from clara.db.models import Transaction

# Acima deste tanto de linhas, atestar em massa esvazia a fila com uma frase
# — e ninguém revisou 500 linhas numa frase só. É o mesmo argumento de escala
# que separa uma correção direta de categoria de `recategorize_transactions`
# em lote. Reabrir nunca exige o cartão: devolver itens à fila é a direção
# segura.
NO_CARD_LIMIT = 20


@dataclass(frozen=True)
class MarkReviewedResult:
    reviewed: int
    reopened: int
    not_found: list[str] = field(default_factory=list)
    audited_by: str = ""
    note: str | None = None


def mark_reviewed(
    session: Session,
    tenant_id: str,
    user_id: str,
    transaction_ids: list[str],
    *,
    reopen: bool = False,
) -> MarkReviewedResult:
    values = (
        {"reviewed_at": None, "reviewed_by": None}
        if reopen
        else {"reviewed_at": datetime.now(UTC), "reviewed_by": f"human:{user_id}"}
    )
    touched = (
        session.execute(
            update(Transaction)
            .where(
                Transaction.tenant_id == tenant_id,
                Transaction.id.in_(transaction_ids),
                # Rascunho não se atesta: ainda vai passar pela decisão do lote
                # inteiro, que é uma conferência mais forte que esta.
                Transaction.status.in_(["confirmed", "adjustment"]),
            )
            .values(**values)
            .returning(Transaction.id)
        )
        .scalars()
        .all()
    )

    done = set(touched)
    missed = [t for t in transaction_ids if t not in done]

    return MarkReviewedResult(
        reviewed=0 if reopen else len(touched),
        reopened=len(touched) if reopen else 0,
        not_found=missed,
        audited_by=f"human:{user_id}",
        note=(
            "Ids em not_found não existem OU ainda estão em lote proposto, onde a revisão é a "
            "decisão do lote."
            if missed
            else None
        ),
    )
